### ISSUE TITLE
Why do we assign DUMMY_VAR__ to containers before calling reader's getter?

### ISSUE THREAD

**Issue by SteveBronder** [stan-dev/stanc3#690]
tl;dr Can we get rid the `fill(Container, DUMMY_VAR__)` when reading in parmeters?

For this example stan model

```stan
data {
  int N;
  int M;
  int K;
  int I[N];
  int J[N];
}

parameters {
  matrix[N, M] A_mat;
}

transformed parameters {
  vector[K] Z_vec = diagonal(A_mat[I, J]);
}
```

In `log_prob` (and `write_array`) we will have code that first creates `A_mat`, then fills it in with the dummy var, then reads in from reader like.

```cpp
    stan::io::reader<local_scalar_t__> in__(params_r__, params_i__);
    local_scalar_t__ DUMMY_VAR__(std::numeric_limits<double>::quiet_NaN());
    (void)DUMMY_VAR__;  // suppress unused var warning

    try {
      Eigen::Matrix<local_scalar_t__, -1, -1> A_mat;
      A_mat = Eigen::Matrix<local_scalar_t__, -1, -1>(N, M);
      stan::math::fill(A_mat, DUMMY_VAR__);

      current_statement__ = 1;
      A_mat = in__.matrix(N, M);
      Eigen::Matrix<local_scalar_t__, -1, 1> Z_vec;
      Z_vec = Eigen::Matrix<local_scalar_t__, -1, 1>(K);
      stan::math::fill(Z_vec, DUMMY_VAR__);

      current_statement__ = 2;
      assign(Z_vec, nil_index_list(),
             diagonal(
                 rvalue(A_mat,
                        cons_list(index_multi(I),
                                  cons_list(index_multi(J), nil_index_list())),
                        "A_mat")),
             "assigning variable Z_vec");
```

But if you look at [reader.hpp](`https://github.com/stan-dev/stan/blob/develop/src/stan/io/reader.hpp`) we are always guaranteed to get back a size `N x M` matrix else it throws an error. For `Z_vec` in the example above (transformed parameters) It makes sense to fill it with nan in case some bad indexing happens. But for paramers (or anything that directly reads in from a `reader` can we get rid of that assign and fill so it's just

```cpp
      current_statement__ = 1;
      Eigen::Matrix<local_scalar_t__, -1, -1> A_mat = in__.matrix(N, M);
```

I'm also working on some stuff right now where, once we can pass around general eigen types, reader.hpp can pass back an `Eigen::Map<>` so we could have

```cpp
      current_statement__ = 1;
      auto A_mat = in__.matrix(N, M);
```

**Comment by bob-carpenter** [comment]
Yes, definitely something that can be optimized. One way I've considered is to have a NaN variable at the bottom of the stack that could be reused.  But in this case, it doesn't even need to be pre-filled.  Eigen's OK with having just garbage memory, whereas `std::vector` requires a proper object on construction.

**Comment by nhuurre** [comment]
The fills were added in #493 . IIRC the problem is that the MIR does not distinguish between parameters and local variables so the code gen cannot know which statements need fill and which do not.
There's also the related issue #503 which again runs into MIR limitations; declarations and assignments are separate statements in the intermediate representation.

**Comment by SteveBronder** [comment]
Ah icic. Is there a way to tag variables in the MIR for whether or not they were generated in a particular section?

**Comment by seantalts** [comment]
For the immediate issue - it would probably not be unreasonable to add a variant to the decl record indicating if it's a data input, parameter input, or local variable here: https://github.com/stan-dev/stanc3/blob/master/src/middle/Stmt.ml#L23

Larger scale, it would be awesome to have a lower intermediate representation MUCH closer to C++ that just the stan_math_backend uses. This would GREATLY simplify the really gnarly code generation process and enable us to do an even larger class of optimizations, e.g. we could remove initializers for any variable that is wholly written to before being read.

**Comment by SteveBronder** [comment]
> For the immediate issue - it would probably not be unreasonable to add a variant to the decl record indicating if it's a data input, parameter input, or local variable here: https://github.com/stan-dev/stanc3/blob/master/src/middle/Stmt.ml#L23

Neat! I'll put this on my to-do list

> Larger scale, it would be awesome to have a lower intermediate representation MUCH closer to C++ that just the stan_math_backend uses. This would GREATLY simplify the really gnarly code generation process and enable us to do an even larger class of optimizations, e.g. we could remove initializers for any variable that is wholly written to before being read.

Like an MIR for the MIR (MMIR?)?  :-P That's probably too much of a big boy job for me

**Comment by seantalts** [comment]
Haha. I believe in you! But it is a much bigger project. If we follow Rust, we'd call it the LIR for Lower Intermediate Representation. They also have a HIR that is distinct from the AST I believe.

**Comment by WardBrian** [comment]
https://github.com/stan-dev/stanc3/pull/930
